components/support/nimbus-fml/src/intermediate_representation.rs

```diff
@@ -54,6 +54,140 @@ pub struct FeatureManifest {
+        let mut enum_names = HashSet::new();
+        self.validate_enum_defs(&mut enum_names)?;
+        // We then validate that each obj_defs also has a unique name.
+        let mut obj_names = HashSet::new();
+        self.validate_obj_defs(&mut obj_names)?;
+
+        // We then validate that each feature_def has a unique name.
+        let mut feature_names = HashSet::new();
+        self.validate_feature_defs(&mut feature_names)?;
+
+        // We then validate that each type_ref is valid
+        for feature_def in &self.feature_defs {
+            for prop in &feature_def.props {
+                self.validate_type_ref(&prop.typ, &enum_names, &obj_names)?;
+            }
+        }
+        self.validate_defaults()?;
+        Ok(())
+    }
+
+    fn validate_type_ref(
+        &self,
+        type_ref: &TypeRef,
+        enum_names: &HashSet<String>,
+        obj_names: &HashSet<String>,
+    ) -> Result<()> {
+        match type_ref {
+            TypeRef::Enum(name) => {
+                if !enum_names.contains(name) {
+                    return Err(FMLError::ValidationError(format!(
+                        "Found enum reference with name: {}, but no definition",
+                        name
+                    )));
+                }
+                Ok(())
+            }
+            TypeRef::Object(name) => {
+                if !obj_names.contains(name) {
+                    return Err(FMLError::ValidationError(format!(
+                        "Found object reference with name: {}, but no definition",
+                        name
+                    )));
+                }
+                Ok(())
+            }
+            TypeRef::EnumMap(key_type, value_type) => {
+                if let TypeRef::Enum(_) = key_type.as_ref() {
+                    self.validate_type_ref(key_type, enum_names, obj_names)?;
+                } else {
+                    return Err(FMLError::ValidationError(format!(
+                        "EnumMap key has be an enum, found: {:?}",
+                        key_type
+                    )));
+                }
+                self.validate_type_ref(value_type, enum_names, obj_names)
+            }
+            TypeRef::List(list_type) => self.validate_type_ref(list_type, enum_names, obj_names),
+            TypeRef::StringMap(value_type) => {
+                self.validate_type_ref(value_type, enum_names, obj_names)
+            }
+            TypeRef::Option(option_type) => {
+                if let TypeRef::Option(_) = option_type.as_ref() {
+                    return Err(FMLError::ValidationError(
+                        "Found nested optional types".into(),
+                    ));
+                }
+                Ok(())
+            }
+            _ => Ok(()),
+        }
+    }
+
+    fn validate_enum_defs(&self, enum_names: &mut HashSet<String>) -> Result<()> {
+        for enum_def in &self.enum_defs {
+            if !enum_names.insert(enum_def.name.clone()) {
```

Comment:
Why try the insert? Why not a simpler contains?

Reply:
uh, it's so that I insert into the set and check if something exists in one function call 😄